net/net_evbuffer.c

```diff
@@ -1551,30 +1564,47 @@ static int process_payload(struct event_info *e)
     }
 }
 
+static void evbuffer_free_callback(const void *ptr, size_t unused0, void *unused1)
+{
+    free((void *)ptr);
+}
+
 static ssize_t readv_evbuffer(struct evbuffer *buf, int fd)
 {
-#   define NVEC 8
-    struct iovec v[NVEC];
-    const int nv = evbuffer_reserve_space(buf, MB(4), v, NVEC);
-    if (nv <= 0) {
+    /*
+       The goal is to read as much as possible, up to 4MB. Previously we used
+       evbuffer_reserve_space()/evbuffer_commit_space(). evbuffer_reserve_space() allocates
+       memory when needed, and links memory chunks in chains; evbuffer_commit_space()
+       pins the chains that data is read into.
+
+       On a congested network this approach ends up wasting a lot memory.
+       A network message may take quite a few events to read completely (a blob operation, for instance).
+       The 1st event asks for 4MB, gets 4MB, and reads only a couple hundred of bytes into
+       the 4MB chain. The 2nd event asks for another 4MB. It can't use the memory from the
+       previous chain because it's pinned. So it gets another 4MB, reads another couple
+       hundred bytes, and pins the entire chain again. And so on so forth...
+
+       Instead, we simly allocate a 4MB chunk of memory and read bytes into the memory.
+       We realloc the memory to the actual number of bytes read. Our malloc implementation
+       guarantees that there's no copy/reposition as long as realloc doesn't expand the size.
+       We then add the memory to the evbuffer using evbuffer_add_reference() so we don't copy
+       the memory again.
+     */
+
+    void *p = malloc(MB(4));
+    if (p == NULL) {
         errno = ENOMEM;
         return -1;
     }
-    const ssize_t sz = readv(fd, v, nv);
+    const ssize_t sz = read(fd, p, MB(4));
     if (sz <= 0) {
         if (errno == EAGAIN || errno == EWOULDBLOCK) return 0;
         return -1;
     }
-    ssize_t n = sz;
-    for (int i = 0; i < nv; ++i) {
-        if (v[i].iov_len < n) {
-            n -= v[i].iov_len;
-            continue;
-        }
-        v[i].iov_len = n;
-        n = 0;
-    }
-    evbuffer_commit_space(buf, v, nv);
+    /* can't possibly fail here */
+    p = realloc(p, sz);
+    if (evbuffer_add_reference(buf, p, sz, evbuffer_free_callback, NULL) != 0)
```

Comment:
Need to think about this a bit - reader has been malloc-heavy and this is doubling the malloc calls for every `read`.

Reply:
Yes - I modified the code a bit:

We get the number of bytes that are immediately available for reading in the socket, via ioctl(FIONREAD). We then do a reserve-readv-commit. If we don't know the info (eg FIONREAD isn't supported or ioctl fails, but I'm guessing this is quite uncommon), we fall back to malloc-read-realloc.

Thoughts?

Reply:
evbuffer_read() [internally does the same](https://github.com/libevent/libevent/blob/master/buffer.c#L2325) per read. So I guess the cost isn't too bad.

Reply:
Yes, but we never called it - we called `readv` directly.

OK, Let us assume for now that additional syscall is not too bad and use `FIONREAD` to set `avail`.  No need for handling the 'bad case'. That's just code which will never run for all practical purposes.